HIP/hip-980.md

```diff
@@ -0,0 +1,114 @@
+
+## User stories
+
+- As a user of mirror node, I want to be able to query historical data for accounts, contracts, tokens, and transactions that were created at a specific timestamp.
+- As a user of mirror node, I want to be able to run locally an `eth_call`, `eth_estimateGas` or `eth_debugTraceTransaction` call that fails on a production environment against the corresponding public state.
+- As a developer of mirror node, I want to be able to test new feature locally against a real life data from a public environment, so that I have better understanding of the feature behavior.
+
+## Specification
+
+The proposed enhancement involves adding a query parameter `timestamp` to the existing routes:
+
+- `/api/v1/accounts/{idOrAliasOrEvmAddress}/tokens` to return correct historical information about an account and their associated fungible tokens for a specific timestamp
+- `/api/v1/accounts/{idOrAliasOrEvmAddress}/nfts` to return correct historical information about an account and their associated nfts for a specific timestamp
+- `/api/v1/accounts/{idOrAliasOrEvmAddress}/allowances/crypto` to return correct historical information about an account's crypto allowances for a specific timestamp
+- `/api/v1/accounts/{idOrAliasOrEvmAddress}/allowances/nfts` to return correct historical information about an account's NFT allowances for a specific timestamp
+- `/api/v1/accounts/{idOrAliasOrEvmAddress}/allowances/tokens` to return correct historical information about an account's token allowances for a specific timestamp
+- `/api/v1/tokens/{tokenId}/nfts/{serialNumber}` to return correct historical information about an NFT for a specific timestamp
+
+**Timestamp Query Parameter**
+
+A new `timestamp` query parameter will be added to search for different resources. This query parameter will represent a consensus timestamp in a Unix format in `seconds.nanoseconds` format and will have an optional comparison operator. That is, it will search for results that `equal`, `not equal`, are `greater` or `less` than the timestamp itself or its seconds component.
+
+`/api/v1/accounts/{idOrAliasOrEvmAddress}/tokens?timestamp={value}`
+`/api/v1/accounts/{idOrAliasOrEvmAddress}/nfts?timestamp={value}`
+`/api/v1/accounts/{idOrAliasOrEvmAddress}/allowances/crypto?timestamp={value}`
+`/api/v1/accounts/{idOrAliasOrEvmAddress}/allowances/nfts?timestamp={value}`
+`/api/v1/accounts/{idOrAliasOrEvmAddress}/allowances/tokens?timestamp={value}`
+`/api/v1/tokens/{tokenId}/nfts/{serialNumber}?timestamp={value}`
+
+**Request**
+```
+/api/v1/accounts/{idOrAliasOrEvmAddress}/tokens?timestamp={value}
+```
+**Example Request**
+
+```
+GET api/v1/accounts/4408244/tokens?order=desc?timestamp=1705334449.803393003
+```
+Note: if ```limit``` is not specified in the request, response will be limited to 25 by default
+
+**Response**
+
+```json
+{
+  "tokens": [
+    {
+      "automatic_association": true,
+      "balance": 1,
+      "created_timestamp": "1705334449.803393003",
```

Comment:
This makes it seem like the user has to know the created_timestamp to pass via the timestamp parameter. We should note somewhere the behavior of the timestamp parameter and how it is linked to the implicit or explicit order. That is we return the state of that entity on or before the given timestamp. It doesn't have to be an exact timestamp match for when it was created or updated. So if order is ascending and `timestamp=X`, `timestamp=eq:X`, or `timestamp=lte:X` the behavior is the same. And vice versa.

Reply:
Added more explanation